### PR TITLE
ya-procachal-tvoi-search-chyvak

### PR DESCRIPTION
**Pull request by tikitko**

### CODE REVIEW

**Comment by YoshkiMatryoshki** [comment]
Any Searches comparisons? Is new version really more optimal?

**Comment by YoshkiMatryoshki** [comment]
How existing post will be updated?

**Review by YoshkiMatryoshki — CHANGES_REQUESTED** [review]
migration. Old posts content will be excluded from search until thay manually updated?

**Comment by tikitko** [comment]
> Any Searches comparisons? Is new version really more optimal?


https://www.postgresql.org/docs/10/textsearch-intro.html

```
Textual search operators have existed in databases for years. PostgreSQL has ~, ~*, LIKE, and ILIKE operators for textual data types, but they lack many essential properties required by modern information systems:

There is no linguistic support, even for English. Regular expressions are not sufficient because they cannot easily handle derived words, e.g., satisfies and satisfy. You might miss documents that contain satisfies, although you probably would like to find them when searching for satisfy. It is possible to use OR to search for multiple derived forms, but this is tedious and error-prone (some words can have several thousand derivatives).

They provide no ordering (ranking) of search results, which makes them ineffective when thousands of matching documents are found.

They tend to be slow because there is no index support, so they must process all documents for every search.
```

**Comment by tikitko** [comment]
> How existing post will be updated?

> migration. Old posts content will be excluded from search until thay manually updated?

not a low of posts now, can be updated manually.

File: blog-server-api/src/main.rs [hunk 1]
```diff
@@ -34,13 +34,41 @@ async fn main() -> screw_components::dyn_result::DResult<()> {
...
+    migrate_db(&rb).await.expect("DB migration failed");
     return rb;
 }
 
+async fn migrate_db(rb: &rbatis::RBatis) -> Result<(), Box<dyn std::error::Error>> {
+    let sql = std::fs::read_to_string("./table_pg.sql")?;
+    rb.exec(&sql, vec![]).await?;
+
+    let posts: Vec<blog_server_services::traits::post_service::Post> =
+        rb.query_decode("select * from post", vec![]).await?;
```
Comment (YoshkiMatryoshki) [inline blog-server-api/src/main.rs:48]: where plain_text_content = NULL?

File: blog-server-api/src/main.rs [hunk 2]
```diff
@@ -34,13 +34,41 @@ async fn main() -> screw_components::dyn_result::DResult<()> {
...
+        .expect("DB init failed");
+    migrate_db(&rb).await.expect("DB migration failed");
     return rb;
 }
 
+async fn migrate_db(rb: &rbatis::RBatis) -> Result<(), Box<dyn std::error::Error>> {
+    let sql = std::fs::read_to_string("./table_pg.sql")?;
+    rb.exec(&sql, vec![]).await?;
+
+    let posts: Vec<blog_server_services::traits::post_service::Post> =
```
Comment (YoshkiMatryoshki) [inline blog-server-api/src/main.rs:47]: We also can add some sort of flag to run this only once

**Review by YoshkiMatryoshki — CHANGES_REQUESTED** [review]

File: blog-server-api/src/main.rs (hunk 1 above)
Comment (tikitko) [reply blog-server-api/src/main.rs]: done

File: blog-server-api/src/main.rs (hunk 2 above)
Comment (tikitko) [reply blog-server-api/src/main.rs]: done

**Review by Bugord — CHANGES_REQUESTED** [review]
Change pull request name

**Comment by tikitko** [comment]
> Change pull request name

image

File: table_pg.sql [hunk 3]
```diff
@@ -222,4 +222,15 @@ DO $$ BEGIN
   END IF;
 END $$
 
+;
+
+DO $$ BEGIN
+  IF NOT EXISTS(select * from pg_tables where schemaname = 'public' and tablename = 'migration') THEN
+    CREATE TABLE migration (
+      key VARCHAR(100) NOT NULL,
+      PRIMARY KEY (key)
```
Comment (YoshkiMatryoshki) [inline table_pg.sql:231]: Would be nice to add date

File: blog-server-api/src/migrations/content.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+pub async fn exec(rb: &rbatis::RBatis) -> Result<(), Box<dyn std::error::Error>> {
+    let is_content_migrated: bool = rb
+        .query_decode::<u64>(
+            "select count(1) as count from migration where key='content_migration'",
```
Comment (YoshkiMatryoshki) [inline blog-server-api/src/migrations/content.rs:4]: (optional) more detailed key? post_content_to_plain/ post_content

**Review by YoshkiMatryoshki — APPROVED** [review]

File: table_pg.sql (hunk 3 above)
Comment (tikitko) [reply table_pg.sql:231]: What reason

Comment (tikitko) [reply table_pg.sql:231]: I also think about renaming of table 'migration' to 'property' with columns 'key string not null' 'value string not null'

Comment (YoshkiMatryoshki) [reply table_pg.sql:231]: to be precise this table is migration__history_ so being able to see in which order updates are executed would be nice

Comment (YoshkiMatryoshki) [reply table_pg.sql:231]: "I'm also thinking about renaming..."
why?

Comment (tikitko) [reply table_pg.sql:232]: storage of properties like as feature flags or for other purposes,
maybe already have plans for this...

File: blog-server-api/src/migrations/content.rs (hunk 4 above)
Comment (tikitko) [reply blog-server-api/src/migrations/content.rs]: done

File: table_pg.sql (hunk 3 above)
Comment (tikitko) [reply table_pg.sql:232]: done

**Pull request merged**